Lib/idlelib/idle_test/test_debugger.py

```diff
@@ -23,7 +34,456 @@ def test_init(self):
+
+    def test_in_rpc_code_rpc_py(self):
+        """Test that .in_rpc_code detects position of rpc.py."""
+
+        # Create a test code object to simulate a debug session.
+        code_obj = compile(TEST_CODE,
+                           'rpc.py',
+                           mode='exec')
+
+        # Create 1 test frame
+        test_frame = MockFrameType(code_obj, 1)
+
+        gui = mock.Mock()
+        gui.interaction = mock.Mock()
+
+        idb = debugger.Idb(gui)
+
+        self.assertTrue(idb.in_rpc_code(test_frame))
+
+    def test_in_rpc_code_debugger_star_dot_py(self):
+        """Test that .in_rpc_code detects position of idlelib/debugger*.py."""
+
+        # Create a test code object to simulate a debug session.
+        for filename in ('idlelib/debugger.py', 'idlelib/debugger_r.py'):
+
+            code_obj = compile(TEST_CODE,
+                               filename,
+                               mode='exec')
+
+            # Create 2 test frames
+            test_frame = MockFrameType(code_obj, 1)
+
+            test_frame2 = MockFrameType(code_obj, 2)
+            test_frame2.f_back = test_frame
+
+            gui = mock.Mock()
+            gui.interaction = mock.Mock()
+
+            idb = debugger.Idb(gui)
+
+            self.assertFalse(idb.in_rpc_code(test_frame2))
+
+
+def make_pyshell_mock():
+    """Factory for generating test fixtures of PyShell."""
+    pyshell = mock.Mock()
+    pyshell.root = None
+    return pyshell
+
+
+class DebuggerTest(unittest.TestCase):
+    """Tests for the idlelib.debugger.Debugger class."""
+
+    def test_setup_debugger(self):
+        """Test that Debugger can be instantiated with a mock PyShell."""
+        pyshell = make_pyshell_mock()
+        test_debugger = debugger.Debugger(pyshell)
+
+        self.assertEquals(test_debugger.pyshell, pyshell)
+        self.assertIsNone(test_debugger.frame)
+
+    def test_run_debugger_with_idb(self):
+        """Test Debugger.run() with an Idb instance."""
+        mock_idb = mock.Mock()  # Mocked debugger.Idb
+        test_debugger = debugger.Debugger(make_pyshell_mock(), idb=mock_idb)
+        test_debugger.run(1, 'two')
+        mock_idb.run.assert_called_once()
+        mock_idb.run.called_with(1, 'two')
+        self.assertEquals(test_debugger.interacting, 0)
+
+    def test_run_debugger_no_idb(self):
+        """Test Debugger.run() with no Idb instance."""
+        test_debugger = debugger.Debugger(make_pyshell_mock(), idb=None)
+        self.assertIsNotNone(test_debugger.idb)
+        test_debugger.idb.run = mock.Mock()
+        test_debugger.run(1, 'two')
+        test_debugger.idb.run.assert_called_once()
+        test_debugger.idb.run.called_with(1, 'two')
+        self.assertEquals(test_debugger.interacting, 0)
+
+    def test_close(self):
+        """Test closing the window in an idle state."""
+        pyshell = make_pyshell_mock()
+        test_debugger = debugger.Debugger(pyshell)
```

Comment:
Since this test and most of the others use the same setup, 
```
        pyshell = make_pyshell_mock()
        test_debugger = debugger.Debugger(pyshell)
```
maybe have this in `setUp()` and `tearDown()`?

I see that some tests instantiate with an idb value and some don't, so maybe have two test runner classes to split those out?  That way, you wouldn't need to repeat the instantiation calls and the cleanup will be in the `tearDown()`.

Reply:
after the `setup` and `teardown`, this function was no longer needed.